src/ray/object_manager/push_manager.cc

```diff
@@ -29,81 +28,91 @@ void PushManager::StartPush(const NodeID &dest_id,
+  auto &dest_map = push_state_map_[dest_id];
+  auto it = dest_map.find(obj_id);
+  if (it == dest_map.end()) {
     chunks_remaining_ += num_chunks;
-    auto push_state = std::make_unique<PushState>(num_chunks, send_chunk_fn);
-    push_requests_with_chunks_to_send_.push_back(
-        std::make_pair(push_id, push_state.get()));
-    push_info_[push_id] = std::move(push_state);
+    dest_map[obj_id] = push_requests_with_chunks_to_send_.emplace(
+        push_requests_with_chunks_to_send_.end(),
+        dest_id,
+        obj_id,
+        num_chunks,
+        std::move(send_chunk_fn));
   } else {
     RAY_LOG(DEBUG) << "Duplicate push request " << push_id.first << ", " << push_id.second
                    << ", resending all the chunks.";
-    if (it->second->NoChunksToSend()) {
-      // if all the chunks have been sent, the push request needs to be re-added to
-      // `push_requests_with_chunks_to_send_`.
-      push_requests_with_chunks_to_send_.push_back(
-          std::make_pair(push_id, it->second.get()));
-    }
-    chunks_remaining_ += it->second->ResendAllChunks(send_chunk_fn);
+    chunks_remaining_ += it->second->ResendAllChunks(std::move(send_chunk_fn));
   }
   ScheduleRemainingPushes();
 }
 
-void PushManager::OnChunkComplete(const NodeID &dest_id, const ObjectID &obj_id) {
-  auto push_id = std::make_pair(dest_id, obj_id);
+void PushManager::OnChunkComplete() {
   chunks_in_flight_ -= 1;
   chunks_remaining_ -= 1;
-  push_info_[push_id]->OnChunkComplete();
-  if (push_info_[push_id]->AllChunksComplete()) {
-    push_info_.erase(push_id);
-    RAY_LOG(DEBUG) << "Push for " << push_id.first << ", " << push_id.second
-                   << " completed, remaining: " << NumPushesInFlight();
-  }
   ScheduleRemainingPushes();
 }
 
 void PushManager::ScheduleRemainingPushes() {
-  bool keep_looping = true;
-  // Loop over all active pushes for approximate round-robin prioritization.
   // TODO(ekl) this isn't the best implementation of round robin, we should
   // consider tracking the number of chunks active per-push and balancing those.
+  // TODO(dayshah): Does round-robin even make sense here? We should probably finish
+  // pushes in the order they were asked for, so that some finish and some work can start.
+  // Otherwise all work will be halted for a period of time
+
+  // Loop over all active pushes for approximate round-robin prioritization.
+  bool keep_looping = true;
   while (chunks_in_flight_ < max_chunks_in_flight_ && keep_looping) {
     // Loop over each active push and try to send another chunk.
-    auto it = push_requests_with_chunks_to_send_.begin();
+    // If we could push out a chunk and haven't reached the chunks_in_flight_ limit,
+    // we'll loop again to try to send more chunks.
     keep_looping = false;
-    while (it != push_requests_with_chunks_to_send_.end() &&
+    auto iter = push_requests_with_chunks_to_send_.begin();
+    while (iter != push_requests_with_chunks_to_send_.end() &&
```

Comment:
I'd drop a comment here that pushes that should be aborted for any reason (e.g., node lost) will delete their entries from the map to exit early